### ISSUE TITLE
Unix paths with backslashes are converted to forward slashes when uri is created with the file scheme

### ISSUE THREAD

**Issue by tmds** [dotnet/corefx#20628]
A backslash is a valid character in a unix path. It is not used to separate path segments, it is part of a directory/file name.
The Uri class will convert these backslashes to forward slashes, causing the resulting uri to no longer point at the correct file.
e.g. `new Uri(file:///one\two)` refers to the `/one/two` instead of `/one\two`

When creating a Uri without the file scheme, the Uri class will preserve the backslashes.
e.g. `new Uri(/one\two)` refers to `/one\two`
Note that the `ToString` of this uri returns `file:///one\two`. Which gives the strange behavior that `new Uri(new Uri(/one\two).ToString())` refers to `/one/two`

The Uri class is very persistent in converting backslashes to forward slashes. Even when providing a percent encoded backslash, it is converted to a forward slash.

One way to fix this is to no longer convert backslashes for `file:///` uris on Unix.
The Uri class has special semantics for two types of files: Dos paths and Unc paths. Dos paths (e.g. c:\) don't start with a slash. Unc paths start with 2 slashes.
So paths starting with a single slash can be treated as unix uris (i.e. no forward slash conversion). This is implemented https://github.com/dotnet/corefx/pull/20623. The behavior is only changed on Unix for paths that start with file:/// (3 slashes = 2 slashes of the scheme + 1 slash of the unix path).

With the current behavior it is not possible to create a Uri with a file:// argument to refer to a path that contains backslashes.
When this behavior is changed, existing code that relied on the backslash-to-forwardslash conversion will break.
An example of this can be seen in this PR (https://github.com/dotnet/corefx/pull/20564). Here the implementation was changed from `new Uri("file://" + path)` to `new Uri(path)`. Due to slashes no longer being converted, some tests needed changes.

@karelz @stephentoub @sepidehMS

**Comment by CIPop** [comment]
The change appears to be related to https://tools.ietf.org/html/rfc8089. 

@tmds are there other RFCs that you referenced and we should investigate prior to deciding on this behavior enhancement?

**Comment by tmds** [comment]
@davidsh Perhaps this should be considered for 2.0, since adding support will break code.
@CIPop From reading the RFCs I think backslashes should be percent encoded en properly restored. The Uri class convert percent encoded backwards slashes to forward slashes. Off course, we must take backwards compatibility into account.

**Comment by tmds** [comment]
@karelz @davidsh This is now future/enhancement. Are those the proper labels? This is a bug and I doubt it will be fixed post 2.0 due to backwards compat.

**Comment by stephentoub** [comment]
@karelz, I agree with @tmds this should be considered for 2.0.

**Comment by tmds** [comment]
@karelz @davidsh ping

**Comment by karelz** [comment]
Switching to 2.0 bug to look into it deeper (I need to read it properly).

**Comment by tmds** [comment]
@karelz @davidsh @CIPop who can take a look at this?

**Comment by karelz** [comment]
@Priya91 is back :) ... @Priya91 can you please help out here?

**Comment by Priya91** [comment]
Sure, looking at this now.

**Comment by Priya91** [comment]
@tmds From the [rfc8089](https://tools.ietf.org/html/rfc8089#appendix-E.4), it specifies the backslashes can be converted to forward slashes if known with reasonable certainty that the backslashes are used as path separators. And since this code is a direct port from Desktop, this behavior is default to cater to Windows. 

So as you say, for Unix posix type paths only, the behavior can be changed to not convert backslashes to forward slashes instead percent encode, without breaking app compat for windows. 

Verified with your initial summary that the bug exists only when passing file:// syntax, without it, the backslashes are percent encoded on Unix. That is,

```c#
string path1 = @"file:///one\two";
string path2 = @"/one\two";
Console.WriteLine($"{new Uri(path1).LocalPath}, {new Uri(path2).LocalPath}");
```

outputs,

```
/one/two, /one%5Ctwo
```

Given I understood the above correctly, you have mentioned that you're fixing this by looking for ```file:///``` syntax, with ```file://``` referring to scheme and ```/``` referring to indication of file path start. But the bug exists for cases where host is specified as well,

```c#
string path1 = @"file://localhost/one\two";
Console.WriteLine($"AbsolutePath: {new Uri(path1).AbsolutePath}");
Console.WriteLine($"LocalPath: {new Uri(path1).LocalPath}");
Console.WriteLine($"ToString: {new Uri(path1)}");
```

outputs,

```
AbsolutePath: /one/two
LocalPath: \\localhost\one\two
ToString: file://localhost/one/two
```

+1 for making the change to support this scenario for all possible test cases on Unix, without breaking Windows behavior.

**Comment by karelz** [comment]
Thanks @Priya91 can you please create the change? Maybe we should run it by shiproom for bar check ... the bar is quite high now :(

**Comment by tmds** [comment]
@Priya91 

The RFC states: 
> It may be possible to translate or update such an invalid file URI by
   replacing all backslashes "\" with slashes "/" if it can be
   determined with reasonable certainty that the backslashes are
   intended as path separators.

This assumption is reasonable on Windows.

What doesn't make sense is the conversion of percent encoded backslashes to forward slashes. Unfortunately, this is what the Uri class is doing.
e.g. `new Uri(@"file:///one%5Ctwo").LocalPath` is `/one/two`

What OS and dotnet version did you use for running the code.
On Linux using 2.0.0-preview1-005977 I get something different (which matches with my expectation)
```
/one/two, /one\two
```
As a background read, this is a nice description of how file URIs should be encoded on the Linux desktop: http://equinox-project.org/spec/file-uri-spec.txt.

@karelz This change is implemented in: https://github.com/dotnet/corefx/pull/20623. @Priya91 can you take a look?

**Comment by Priya91** [comment]
> What OS and dotnet version did you use for running the code.

On ubuntu, with latest master bits.

**Comment by tmds** [comment]
> This is expected, since backslash is not a reserved character in URI, and does not require to be percent-encoded. I don't think this is a bug. This happens in code here

You misread my comment. `%5C` is a backslash. So it is strange `new Uri(@"file:///one%5Ctwo").LocalPath` turns it into a forward slash.

**Comment by Priya91** [comment]
@tmds I did understand it correctly, what I meant was it's not a new bug, it's the same case that '\' is converted to '/'. The '\' being percent-encoded doesn't protect it from the conversion rule.

**Comment by tmds** [comment]
> it's the same case that '\' is converted to '/'.

FWIW: the RFC section that allows converting backward slashes to forward slashes only applies to unencoded slashes.

**Comment by tmds** [comment]
@Priya91 The LocalPath should not contain the encoded backslash (output you observed), since this will give issues, e.g when passed as an argument to File.Open.

**Comment by Priya91** [comment]
> The LocalPath should not contain the encoded backslash (output you observed),

That's for the case when file scheme is not used.

**Comment by tmds** [comment]
> That's for the case when file scheme is not used.

Yes, it would cause bugs if the LocalPath contains percent encoded characters.

When I run the code with the released 2.0.0-preview1 the LocalPath has no encoded characters. I tried using the latest (Microsoft.NETCore.App 2.1.0-preview1-25413-05). Still the app outputs `/one/two, /one\two`. I can not reproduce the percent encoded backslash in LocalPath.

**Comment by tmds** [comment]
@Priya91 is proposing to extend the PR so backslashes are preserved also for file uris that have a hostname.
e.g. `file://hostname/folder\path` would change the AbsoluteUri from `file://hostname/folder/path` to `file://hostname/folder%5Cpath`.
This seems fine. However, to convert a file uri back into a string the host system can understand (e.g. to pass to `File.Open`), one uses the `LocalPath`.
And this uri is considered a Unc uri. Its `IsUnc` property is `true` and its `LocalPath` is `\\hostname\folder\filepart1\filepart2`.

Should the PR address these Uris too? For a `file://hostname/folder\path` what would be the desired `AbsoluteUri`, `LocalPath`, `IsUnc` values?
I feel inclined to preserve the Unc semantics and not change the behavior, except perhaps for the special case where hostname = localhost? That is: `file://localhost/` is treated the same as `file:///`. This is also considered a special case by http://equinox-project.org/spec/file-uri-spec.txt.

@Priya91 @davidsh @CIPop @stephentoub what do you think?

**Comment by tmds** [comment]
I'd not extend the PR (even not for localhost) and maintain the Unc semantics.
I don't expect there to be code that does `"file://somehost/" + Path.GetFullPath(...)` since that is treated as a Unc uri and would not work.
Where as the case handled by the PR `"file://" + Path.GetFullPath(...)` is a local file path already.

**Comment by Priya91** [comment]
>@Priya91 @davidsh @CIPop @stephentoub what do you think?

Just thinking about it holistically, I would expect the backslashes to not be converted to forwardslashes in unix file uri's path component, since the backslash is a valid char for filenames on Unix, contrary to Windows. And since UNC is a windows concept, I wouldn't put a lot of weight for using that reason to justify not fixing the bug for cases where the file uri's localpath will output an UNC on Unix. I would expect the developers on Unix, to not use LocalPath since that path will be invalid there.

**Comment by tmds** [comment]
@Priya91 can you come up with a piece of code that works on Windows, does not work on Unix, and would be fixed by extending the PR?

**Comment by Priya91** [comment]
@tmds I'm not sure what you are expecting? Are you asking for cases where applying the fix for Unix with hostnames can be a scenario? Accessing a file on another linux box that has a backward slash in its filename is not a common scenario?

**Comment by tmds** [comment]
The scenario is fine. I don't think it has code that would work better by extending the PR.

**Comment by tmds** [comment]
... because the extended PR allows you to put the backslash in the Uri, but there is no convenient way to get back the path. AbsoluteUri is percent encoded, LocalPath returns a Unc path.

**Comment by tmds** [comment]
@karelz @CIPop @Priya91 @stephentoub @davidsh @sepidehMS can you review the PR (https://github.com/dotnet/corefx/pull/20623)? We should decide if this can be in 2.0. If not, it shouldn't be in a later release due to backwards compat.

**Comment by karelz** [comment]
Fixed in 2.0 in PR #21290
